stdlib/public/core/StringStorageBridge.swift

```diff
@@ -335,6 +373,464 @@ extension __SharedStringStorage {
+          return String(decoding: arr, as: UTF16.self)
+        }
+      }
+    }
+    
+    var start: UnsafePointer<UInt8>? {
+      @_effects(readonly) @inline(__always) get {
+        switch self {
+        case .native(let str):
+          let guts = str._guts
+          guard !guts.isSmall else { return nil }
+          return guts._object.fastUTF8.baseAddress!
+        default:
+          return nil
+        }
+      }
+    }
+    
+    var isASCII: Bool {
+      @_effects(readonly) @inline(__always) get {
+        switch self {
+        case .native(let str):
+          return str._guts.isASCII
+        default:
+          return false
+        }
+      }
+    }
+    
+    var UTF16Length: Int {
+      @_effects(readonly) @inline(__always) get {
+        switch self {
+        case .native(let str):
+          return str.utf16.count // UTF16View special-cases ASCII for us.
+        case .utf16(let arr):
+          return arr.count
+        }
+      }
+    }
+    
+    @_effects(readonly)
+    func character(at offset: Int) -> UInt16 {
+      switch self {
+      case .native(let str):
+        return str.utf16[str._toUTF16Index(offset)]
+      case .utf16(let arr):
+        return arr[offset]
+      }
+    }
+    
+    func _validUTF16(_ data: UnsafeBufferPointer<UTF16.CodeUnit>) -> Bool {
+      var lookingForTrailingSurrogate = false
+      for scalar in data {
+        if UTF16.isTrailSurrogate(scalar) {
+          if _fastPath(lookingForTrailingSurrogate) {
+            lookingForTrailingSurrogate = false
+          } else {
+            return false
+          }
+        }
+        if UTF16.isLeadSurrogate(scalar) {
+          if _fastPath(!lookingForTrailingSurrogate) {
+            lookingForTrailingSurrogate = true
+          } else {
+            return false
+          }
+        }
+      }
+      return true
+    }
+    
+    mutating func _switchRepresentationsIfInvalidUTF16(
+      _ replacementBytes: UnsafeBufferPointer<UTF16.CodeUnit>
+    ) {
+      if case .native(let str) = self, !_validUTF16(replacementBytes) {
+        self = .utf16(Array(str.utf16))
+      }
+    }
+    
+    mutating func _convertIncomingNSRangeSwitchingRepresentationIfUnaligned(
+      _ range: _SwiftNSRange
+    ) -> Range<String.Index>? {
+      switch self {
+      case .native(let str):
+        let range = str._toUTF16Indices(
+          range.location ..< range.location + range.length
+        )
+        let rs = range.lowerBound
+        let re = range.upperBound
+        let scalarAligned =
+          (rs._isScalarAligned || rs.transcodedOffset == 0) &&
+            (re._isScalarAligned || re.transcodedOffset == 0)
+        
+        if _slowPath(!scalarAligned) {
+          self = .utf16(Array(str.utf16))
+          return nil
+        }
+        return range
+      case .utf16(_):
+        return nil
+      }
+    }
+    
+    // Fast mutation path: replacement is known valid, we are known String-backed
+    mutating func _replace(
```

Comment:
Should this just be in the body of replace? Or at least always inlined into replace? At the very least, have "fast" or some such in the name so the assertion is clear from the name.